src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs

```diff
@@ -263,25 +251,34 @@ public AnalysisResults(bool ignoreCase, bool allAsciiIfIgnoreCase, int hashIndex
             public bool RightJustifiedSubstring => HashIndex < 0;
         }
 
+        private delegate ReadOnlySpan<char> GetSpan(string s, int index, int count);
+        private static GetSpan FromLeft = [MethodImpl(MethodImplOptions.AggressiveInlining)] static (string s, int index, int count) => s.AsSpan(index, count);
+        private static GetSpan FromRight = [MethodImpl(MethodImplOptions.AggressiveInlining)] static (string s, int index, int count) => s.AsSpan(s.Length + index, count);
+        private static GetSpan FullString = [MethodImpl(MethodImplOptions.AggressiveInlining)] static (string s, int index, int count) => s.AsSpan();
+
         private abstract class SubstringComparer : IEqualityComparer<string>
         {
             public int Index;
             public int Count;
-            public bool IsLeft;
+            public GetSpan Slicer;
+
+            public SubstringComparer(GetSpan slicer) { Slicer = slicer; }
```

Comment:
By requiring the slicer to be provided at construction, we don't have to worry about nulls